daemon,k8s: ServiceCache interface in preparation for adapter

joamaki opened the pull request:
For the transition into the new load-balancing control-plane we will need to adapt the existing uses of ServiceCache and ServiceManager to the new code. Prepare the APIs for writing adapters that bridge the old API to the new control-plane:

* Introduce k8s.ServiceCache interface implemented by ServiceCacheImpl
* 'MinimalService' and 'MinimalEndpoints' in the ServiceCache Notifications() to minimize required conversion work needed.

(pulled out from https://github.com/cilium/cilium/pull/37083)

joamaki committed:
daemon,k8s: ServiceCache interface in preparation for adapter

For the transition into the new load-balancing control-plane
we will need to adapt the existing uses of ServiceCache and
ServiceManager to the new code. Prepare the APIs for writing
adapters that bridge the old API to the new control-plane:

* Introduce k8s.ServiceCache interface implemented by ServiceCacheImpl
* 'MinimalService' and 'MinimalEndpoints' in the ServiceCache Notifications()
  to minimize required conversion work needed.
* Remove unsed MergeClusterService*
* Remove unused GetServiceIP() and GetEndpointsOfService()
* Move LooseMatch to service_cache_test.go as only tests need it

Signed-off-by: Jussi Maki <jussi@isovalent.com>

squeed commented on pkg/k8s/service_cache.go:
```diff
@@ -107,20 +108,113 @@ type ServiceEvent struct {
+	Selector    map[string]string
+}
+
+func (ms *MinimalService) DeepEqual(other *MinimalService) bool {
```
use DeepEqual-gen?

squeed commented on pkg/k8s/service_cache.go:
```diff
@@ -107,20 +108,113 @@ type ServiceEvent struct {
+	return meps
+}
+
+type ServiceCache interface {
```
Random thought: call this ServiceStore? Or is this all going away "soon"?

squeed replied:
Oh, nevermind, ServiceCache is the existing type :coffee:

joamaki replied:
It's going away soon. Or rather it gets removed from cilium-agent and we'll have a minimal version of "ServiceCache" in the operator for at least a bit longer.

squeed commented on pkg/k8s/service_cache.go:
```diff
@@ -107,20 +108,113 @@ type ServiceEvent struct {
+}
+
+type ServiceCache interface {
+	Events() <-chan ServiceEvent
```
suggestion: Events() is not really for general use; only one consumer is allowed. We should call this out in the docblock, and maybe even `Error("BUG...")` if called more than once for the same ServiceCache.

joamaki replied:
hmm, tests call it multiple times. OK if we'll just document this? It's not going to live very long. And in the operator-side use of this we can make that one unexported.

squeed replied:
deal.

squeed replied:
And yes, it's not a good API :-).

joamaki commented on pkg/policy/k8s/service.go:
```diff
@@ -39,12 +39,13 @@ func (p *policyWatcher) onServiceEvent(event k8s.ServiceNotification) {
-func (p *policyWatcher) updateToServicesPolicies(svcID k8s.ServiceID, newSVC, oldSVC *k8s.Service) error {
+func (p *policyWatcher) updateToServicesPolicies(svcID k8s.ServiceID, newSVC, oldSVC *k8s.MinimalService, newEps, oldEps *k8s.MinimalEndpoints) error {
 	var errs []error
+	endpointsChanged := !newEps.DeepEqual(oldEps)
```
This PR might've uncovered an existing bug. We were only checking the service for equality, but we actually also cared about the endpoints changing. I expect this was the cause for this failure: https://github.com/cilium/cilium/actions/runs/13389227688/job/37393038062 (to be checked...). Not sure why exactly this worked before when it was comparing the more complex `*Service` for equality, guess something in there changed when endpoints changed.

squeed replied:
does it make sense to add a zero-sized uncomparable field in the struct? Hopefully no more demons are awoken :see_no_evil:

giorio94 commented on pkg/policy/k8s/service.go:
```diff
@@ -39,12 +39,13 @@ func (p *policyWatcher) onServiceEvent(event k8s.ServiceNotification) {
-func (p *policyWatcher) updateToServicesPolicies(svcID k8s.ServiceID, newSVC, oldSVC *k8s.Service) error {
+func (p *policyWatcher) updateToServicesPolicies(svcID k8s.ServiceID, newSVC, oldSVC *k8s.MinimalService, newEps, oldEps *k8s.MinimalEndpoints) error {
 	var errs []error
+	endpointsChanged := !newEps.DeepEqual(oldEps)
```
Nit. Could be worth moving this fix to a dedicated commit, so that is it easier to backport if necessary.

joamaki replied:
Done.

giorio94 commented on pkg/k8s/service_cache.go:
```diff
@@ -107,20 +107,147 @@ type ServiceEvent struct {
+	}
+}
+
+// MinimalService is a slimmed down version of 'Endpoints'.
```
Nit.

```suggestion
// MinimalEndpoints is a slimmed down version of 'Endpoints'.
```

joamaki committed:
policy/k8s: Also check if endpoints have changed

In updateToServicesPolicies() we are checking if the service
has changed. This isn't enough since we are also processing
the endpoints in resolveToServices() and thus we also need
to check if the endpoints have changed.

Fixes: 1ff9be2d77e2 ("policy/k8s: Move ToServices translation to policy package")
Signed-off-by: Jussi Maki <jussi@isovalent.com>
